lib/active_merchant/billing/gateways/sigma.rb

```diff
@@ -0,0 +1,224 @@
+          .gsub(%r{(<tem:APIKey>).+(</tem:APIKey>)}, '\1[FILTERED]\2')
+          .gsub(%r{(<tem:CVV>).+(</tem:CVV>)}i, '\1[FILTERED]\2')
+      end
+
+      private
+
+      def add_card_info(post, payment, options)
+        post[:cvv] = payment.verification_value
+        post[:accountToken] =
+          options[:card_token].presence || store(payment, options).authorization
+      end
+
+      def add_customer_data(action, post, options)
+        post[:APIKey] = @options[:api_key]
+
+        if action == 'purchase'
+          post[:merchantAccountNumber] = @options[:merchant_id]
+          post[:terminalName] = @options[:terminal_id]
+          post[:emailAddress] = options[:email]
+          post[:accessCode] = options[:access_token] || '123123'
+          post[:clientTracking] = options[:order_id]
+        elsif action == 'store'
+          post[:AccountNumber] = options[:order_id]
+        end
+      end
+
+      def add_additional_data(post, options)
+        if options.key?(:additional_data).present?
+          additional_data = {}
+          additional_data[:name] = options.dig(:additional_data, :name)
+          additional_data[:value] = options.dig(:additional_data, :value)
+          post[:additionalData] = additional_data
+        end
+      end
+
+      def add_invoice(post, money, options)
+        post[:amount] = amount(money)
+        post[:currencyCode] = (options[:currency] || currency(money))
+      end
+
+      def add_payment(post, card)
+        post[:CardholderName] = card.name
+        post[:CardNumber] = card.number
+        post[:CVV] = card.verification_value
+        post[:ExpirationDate] =
+          format(card.month, :two_digits) + format(card.year, :two_digits)
+      end
+
+      def add_shipping_data(sale, options)
+        sale[:shippingName] =
+          options.dig(:address, :name) || options[:full_name]
+        sale[:shippingDate] = options.dig(:address, :date) || ''
+        sale[:shippingAddress] = options.dig(:address, :address1)
+        sale[:shippingCity] = options.dig(:address, :city)
+        sale[:shippingState] = options.dig(:address, :state)
+        sale[:shippingCountry] = options.dig(:address, :country)
+        sale[:shippingZipCode] = options.dig(:address, :zip)
+        sale[:shippingPhoneNumber] = options.dig(:phone)
+      end
+
+      def add_billing_data(sale, options)
+        sale[:billingAddress] =
+          options.dig(:address, :name) || options[:full_name]
+        sale[:billingCity] = options.dig(:address, :city)
+        sale[:billingState] = options.dig(:address, :state)
+        sale[:billingCountry] = options.dig(:address, :country)
+        sale[:billingZipCode] = options.dig(:address, :zip)
+        sale[:billingPhoneNumber] = options.dig(:phone)
+      end
+
+      def parse(body)
+        Nokogiri.XML(body)
+      end
+
+      def headers(data)
+        {
+          'Content-Type' => 'application/soap+xml; charset=utf-8',
+          'Content-Length' => data.size.to_s
```

Comment:
This looks like it should take encoding into account. For example, a string with umlaut `ü` is size 1 but byte size 2:

```
irb(main):013:0> "\u00FC"
=> "ü"
irb(main):014:0> "\u00FC".size
=> 1
irb(main):004:0> "\u00FC".bytes.size
=> 2
```

Reply:
ahh nice trick, thanks!